flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1505,22 +1505,27 @@ class ScalarExprLowering {
     return std::visit([&](const auto &x) { return genval(x); }, dref.u);
   }
 
-  // Helper function to turn the left-recursive Component structure into a list
-  // that does not contain allocatable or pointer components other than the last
-  // one.
-  // Returns the object used as the base coordinate for the component chain.
+  // Helper function to turn the Component structure into a list of nested
+  // components, ordered from largest/leftmost to smallest/rightmost:
+  //  - where only the smallest/rightmost item may be allocatable or a pointer
+  //    (nested allocatable/pointer components require nested coordinate_of ops)
+  //  - that does not contain any parent components
+  //    (the front end places parent components directly in the object)
+  // Return the object used as the base coordinate for the component chain.
   static Fortran::evaluate::DataRef const *
   reverseComponents(const Fortran::evaluate::Component &cmpt,
-                    std::list<const Fortran::evaluate::Component *> &list) {
-    list.push_front(&cmpt);
+                    std::list<const Fortran::evaluate::Component *> &list,
+                    bool isParentComponent = false) {
+    if (!isParentComponent)
+      list.push_front(&cmpt);
     return std::visit(
         Fortran::common::visitors{
             [&](const Fortran::evaluate::Component &x) {
-              // Stop the list when a component is an allocatable or pointer
-              // because the component cannot be lowered into a single
-              // fir.coordinate_of.
-              if (Fortran::semantics::IsAllocatableOrPointer(x.GetLastSymbol()))
+              const Fortran::semantics::Symbol &sym = x.GetLastSymbol();
+              if (Fortran::semantics::IsAllocatableOrPointer(sym))
                 return &cmpt.base();
+              if (sym.detailsIf<Fortran::semantics::DerivedTypeDetails>())
```

Comment:
OK, I am lost here. I would have expected `sym.test(Fortran::semantics::Symbol::Flag::ParentComp)` to be the way to detect parent component, but it crashes your test.

However, your change still crashes if the user did specify the parent component himself (e.g., writing `u3v%u2%u2t3%t1i` explicitly ), and in that case, using `sym.test(Fortran::semantics::Symbol::Flag::ParentComp)` detects this case and solves it.

Reply:
Good catch.  The front end is incorrectly canonicalizing an optional parent component name with a component type symbol rather than the associated same-named component symbol.  I've updated the code to assumed the corrected form.  Peter has the front end fix under review.